### PR TITLE
filesource schema discovery & source-http-file streaming

### PR DESCRIPTION
**Pull request by jgraettinger**
**Description:**

Make `filesource` connector discover work _better_, though still not _great_.

source-http-file supports unbounded real-time streams, such as wikipedia recent changes.

**Workflow steps:**

* (Optional) Use the new "Headers" tab of source-http-file to provide additional headers to be sent with the request.

* Start using source-http-file with streaming data sources, like wikimedia's recent changes, that previously did not work.

**Documentation links affected:**

(list any [documentation links](https://docs.google.com/document/d/1SRC9VS9zyCzWl3n4HXHbc4wPB1eLxJHkA2rtu9ZNokM/edit?usp=sharing) that you created, or existing ones that you've identified as needing updates, along with a brief description)

**Notes for reviewers:**

(anything that might help someone review this PR)

- - -
This change is [Reviewable](https://reviewable.io/reviews/estuary/connectors/321)

### CODE REVIEW

**Comment by jgraettinger** [comment]
Image of headers in UI:
Screen Shot 2022-08-21 at 2 14 24 PM

**Comment by jgraettinger** [comment]
Manual tested with:
 * Wiki recent changes (discovery and capture)
 * source-s3 discovery with `tripdata` citibike bucket.

File: filesource/discovery.go [hunk 1]
```diff
@@ -3,261 +3,194 @@ package filesource
...
-						"type": "integer",
-						"minimum": 0
-					}
-				},
-				"required": ["file", "offset"]
-			}
-		},
-		"required": ["_meta"]
-	}`)
+// The maximum amount of time we'll read documents for.
```
Comment (willdonnelly) [inline filesource/discovery.go:22]: If I'm understanding the code properly it looks like we could actually take up to `2*discoverBudget` for discovery (once for the listing to time out, then another time for the reading). Is that right?.

Perhaps this should be mentioned in the comment, or else have separate `discoverListingBudget` and `discoverReadBudget` constants?

File: filesource/discovery.go [hunk 2]
```diff
@@ -3,261 +3,194 @@ package filesource
...
+		if pathRe, err = regexp.Compile(r); err != nil {
+			return fmt.Errorf("building regex: %w", err)
+		}
 	}
 
-	return schema, nil
-}
+	// Step 1: select up to N objects to read using reservoir sampling.
+	// List until we read EOF or exceed our time budget.
+	var reservoir = make([]ObjectInfo, 0, 25)
```
Comment (willdonnelly) [inline filesource/discovery.go:41]: nit: Pull the magic number `25` out as a named constant?

File: source-http-file/main.go [hunk 3]
```diff
@@ -20,10 +20,18 @@ type credentials struct {
 	Password string `json:"password"`
 }
 
+type headers struct {
+	Items []struct {
```
Comment (willdonnelly) [inline source-http-file/main.go:24]: Maybe call this `Extra` or something? Or just remove the extra level of nesting if you don't anticipate wanting to add `/headers/foo` properties for specific other headers?

**Review by willdonnelly — APPROVED** [review]
LGTM

**Review by jgraettinger — COMMENTED** [review]
Thanks

> *[Reviewable](https://reviewable.io/reviews/estuary/connectors/321#-:-NA6X1iuBGPqdRX_uHR_:b-zcjvte)* status: 0 of 8 files reviewed, 2 unresolved discussions (waiting on @willdonnelly)

___
*[`filesource/discovery.go` line 22 at r2](https://reviewable.io/reviews/estuary/connectors/321#-NA0kNyIn9IsAr1MhnNu-r2-22:-NA6UuUz-bFt_4XMvgTG:b-d8y06e) ([raw file](https://github.com/estuary/connectors/blob/ac9ae369fd26ac5da48c6c5a86cf30459a40f3a8/filesource/discovery.go#L22)):*
<details><summary><i>Previously, willdonnelly (Will Donnelly) wrote…</i></summary><blockquote>

If I'm understanding the code properly it looks like we could actually take up to `2*discoverBudget` for discovery (once for the listing to time out, then another time for the reading). Is that right?.

Perhaps this should be mentioned in the comment, or else have separate `discoverListingBudget` and `discoverReadBudget` constants?
</blockquote></details>

that's right. Updated comment.
___
*[`filesource/discovery.go` line 41 at r2](https://reviewable.io/reviews/estuary/connectors/321#-NA0kNyIn9IsAr1MhnNu-r2-41:-NA6W-vs2wDM_lNE5Tfn:b-896fix) ([raw file](https://github.com/estuary/connectors/blob/ac9ae369fd26ac5da48c6c5a86cf30459a40f3a8/filesource/discovery.go#L41)):*
<details><summary><i>Previously, willdonnelly (Will Donnelly) wrote…</i></summary><blockquote>

nit: Pull the magic number `25` out as a named constant?
</blockquote></details>

Done.
___
*[`source-http-file/main.go` line 24 at r2](https://reviewable.io/reviews/estuary/connectors/321#-NA0kNyIn9IsAr1MhnNy-r2-24:-NA6WaHeCC-I6otGWDIp:b19rpfx) ([raw file](https://github.com/estuary/connectors/blob/ac9ae369fd26ac5da48c6c5a86cf30459a40f3a8/source-http-file/main.go#L24)):*
<details><summary><i>Previously, willdonnelly (Will Donnelly) wrote…</i></summary><blockquote>

Maybe call this `Extra` or something? Or just remove the extra level of nesting if you don't anticipate wanting to add `/headers/foo` properties for specific other headers?
</blockquote></details>

added a comment to clarify why this is nested.

**Pull request merged**